### DISCUSSION TITLE
ES5 e o uso do "use Strict"

### DISCUSSION THREAD

**Discussion by vitocaio** [frontendbr/forum#257] [Perguntas e Respostas]
Ola amigos do Frontbt!

Eu gostaria de colocar em pauta o seguinte assunto. Mesmo com um tempo considerável no mundo do JS ainda não compreendo direito o uso do "use strict", eu sei que é uma forma de acessar tudo que existe de mais novo no mundo do JS, mas como me manter atualizado? onde posso aplicar as novas regras do ES5, etc... Um bom front, deve dominar a linguagem e suas novidades, quero me renovar e entender melhor as peculiaridades da linguagem, obrigado!

**Comment by renatogalvones** [comment]
> eu sei que é uma forma de acessar tudo que existe de mais novo no mundo do JS

Fala @vitocaio! Posso estar fazendo confusão, mas por essa citação acima posso afirmar que **strict mode** não é bem isso não.

Direto do MDN:

> O strict mode do ECMAScript 5 é uma forma de optar por uma variante restrita do JavaScript.
> 
> O strict mode faz várias mudanças nas semânticas normais do JavaScript. Primeiro, o strict mode elimina alguns erros silenciosos do JavaScript fazendo-os lançar excessões. Segundo, o strict mode evita equívocos que dificultam que motores JavaScript realizem otimizações: código strict mode pode às vezes ser feito para executar mais rápido que código idêntico não-strict mode. Terceiro, strict mode proíbe algumas sintaxes que provavelmente serão definidas em versões futuras do ECMAScript.

Dá uma conferida!
https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Strict_mode

E respondendo suas perguntas sobre como se atualizar o @willianjusten fez um post que tem uma lista bacana:
https://willianjusten.com.br/como-se-tornar-um-desenvolvedor-front-end/#update

E tb um post do FrontEndRescue
https://uptodate.frontendrescue.org/pt/

Bons estudos! :metal:

**Comment by vitor-mendrone-deel** [comment]
@vitocaio o "use strict" resolve alguns problemas que muitas vezes passam despercebidos. O principal dele (pelo menos pra mim que muitas vezes escrevo código na correria) é não permitir o uso de variáveis em funções sem serem declaradas previamente, por exemplo: 

Sem `"use strict"`, a variável abaixo declarada dentro da função seria "içada" pra fora e passaria a estar em um escopo global: 

``` javascript
(function() {
    someVar = "variable";
}());
```

Seria interpretado como: 

``` javascript
var someVar;
(function() {
    someVar = "variable";
}());
```

Note que a variável foi "içada" ao escopo global, o que pode ser um problema sério em aplicações maiores. 

Outra coisa bastante útil no `"use strict"` é o fato de não permitir propriedades duplicadas em objetos: 

``` javascript

var myObject = {
    bar: 1,
    bar: 2
}
```

Nesse caso seria gerado um erro. 

Existem outros casos em que o strict mode força o uso de boas práticas e evita que erros ocorram, é legal pesquisar sobre pra entender melhor como e quando e porque aplicar o strict mode. 

Uma última dica: evite usar o `"use strict"` em escopo global pra evitar erros quando concatenar o código (usando minificadores/concatenadores, por exemplo), ou seja: 

Evite isso: 

``` javascript
"use strict";

function myFunction() {

}
```

Prefira isso: 

``` javascript
function myFunction() {
    "use strict";
}
```

**Comment by vitocaio** [comment]
Pow, legal boas dicas, você sabe me dizer se tem algum artigo abordando este assunto em questão?

**Comment by lagden** [comment]
no **backend** sem medo de ser feliz:

``` javascript
"use strict";

function fuckingInHeaven() {
  return true;
}
```

já no **front** é melhor dentro do escopo local... avoid bugs!!!
como explicou o @Mendrone no https://github.com/frontendbr/forum/issues/257#issuecomment-235056897
